Adds function to define camera configs through intrinsic matrix

pascal-roth opened the pull request:
# Description

This PR adds the possibility of initializing cameras (both Raycaster Cameras and USD Cameras) with the intrinsic matrix instead of using the aperture parameters. The intrinsic matrix is defined in the pattern config and the pinhole cameras spawn config, respectively. 

Moreover, it fixes the bug that the vertical aperture is not adjusted for the USD camera case (it will always stay at the default value, even if the horizontal aperture is set). The default is squared pixels; however, it also allows the setting of other values.

Fixes https://github.com/isaac-orbit/IsaacLab/issues/226

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

pascal-roth commented:
@Mayankm96 I remember there were issues when the horizontal and vertical apertures differed for the USD camera case. Do you know if that has been resolved by now? Just from looking at the resulting image, I couldn't see something wrong, but otherwise, I will check again.

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/ray_caster/patterns/patterns_cfg.py:
```diff
@@ -91,6 +91,22 @@ class PinholeCameraPatternCfg(PatternBaseCfg):
     """Offsets Resolution/Film gate horizontally. Defaults to 0.0."""
     vertical_aperture_offset: float = 0.0
     """Offsets Resolution/Film gate vertically. Defaults to 0.0."""
+    intrinsic_matrix: tuple[float] | None = None
+    """Intrinsic matrix of the camera. Defaults to None.
+
+    The intrinsic matrix is a 3x3 matrix that defines the mapping between the 3D world coordinates and the 2D image.
+    The matrix is defined as:
+
+    .. math::
+        \\begin{bmatrix}
+        f_x & 0 & c_x \\\\
+        0 & f_y & c_y \\\\
+        0 & 0 & 1
+        \\end{bmatrix}
+
+    Note:
+        In the case both, horizontal aperture and intrinsic matrix is defined, the intrinsic matrix will be used.
+    """
```
This will then reduce the changes in other places and make it nicely contained just in the config class :)

Mayankm96 commented on source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets/stereolabs.py:
```diff
@@ -0,0 +1,63 @@
+# Configuration as RayCasterCameraCfg
+##
+
+ZED_X_NARROW = RayCasterCameraCfg(
```
Can we rename these to be the type of camera implementation as well? Something like `ZED_X_NARROW_RAYCASTER_CFG`? Also we end all configs with `CFG`.

Mayankm96 commented on source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets/stereolabs.py:
```diff
@@ -0,0 +1,63 @@
+    debug_vis=True,
+    max_distance=20,
+    data_types=["distance_to_image_plane"],
+)
```
Would be great to add proper docs on this with links to the actual camera (similar to how we do for the robots.

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/camera/camera.py:
```diff
@@ -117,6 +116,18 @@ def __init__(self, cfg: CameraCfg):
             rot_offset = rot_offset.squeeze(0).numpy()
+            # if intrinsic matrix is set, override aperture parameters in config
+            if self.cfg.spawn.intrinsic_matrix is not None:
+                # extract parameters from matrix
+                f_x = self.cfg.spawn.intrinsic_matrix[0]
+                c_x = self.cfg.spawn.intrinsic_matrix[2]
+                f_y = self.cfg.spawn.intrinsic_matrix[4]
+                c_y = self.cfg.spawn.intrinsic_matrix[5]
+                # resolve parameters for usd camera
+                self.cfg.spawn.horizontal_aperture = self.cfg.width * self.cfg.spawn.focal_length / f_x
+                self.cfg.spawn.vertical_aperture = self.cfg.height * self.cfg.spawn.focal_length / f_y
+                self.cfg.spawn.horizontal_aperture_offset = (c_x - self.cfg.width / 2) / f_x
+                self.cfg.spawn.vertical_aperture_offset = (c_y - self.cfg.height / 2) / f_y
```
Can we please have a unit test that checks that intrinsic matrix is loaded correctly and images are exactly what we expect (from both intrinsic matrix and direct USD settings)?

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/camera/camera.py:
```diff
@@ -541,17 +540,21 @@ def _update_intrinsic_matrices(self, env_ids: Sequence[int]):
             # get camera parameters
             focal_length = sensor_prim.GetFocalLengthAttr().Get()
             horiz_aperture = sensor_prim.GetHorizontalApertureAttr().Get()
+            vert_aperture = sensor_prim.GetVerticalApertureAttr().Get()
+            horiz_aperture_offset = sensor_prim.GetHorizontalApertureOffsetAttr().Get()
+            vert_aperture_offset = sensor_prim.GetVerticalApertureOffsetAttr().Get()
```
Do these now work properly from rendering PoV? I don't know if this issue is resolved from Omniverse side. Also vertical and horizontal aperture should be same for "square" pixels?

pascal-roth replied:
I don't know if they work properly. Will implement a test to see if the results between between raycaster and USD camera are the same if I set them

pascal-roth replied:
apparently it is not resolved

pascal-roth replied:
offesets are not resolved

jtigue-bdai replied:
From what I remember the vertical aperture doesn't actually change the rendering of the image. Ultimately it comes down to the resolutions and horizontal aperture. The vertical aperture does change the visualized frustum though.

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/ray_caster/patterns/patterns_cfg.py:
```diff
@@ -96,6 +96,53 @@ class PinholeCameraPatternCfg(PatternBaseCfg):
+
+        Args:
+            focal_length: Focal length of the camera (in cm).
+            intrinsic_matrix: Intrinsic matrix of the camera (shape: (9,)).
```
Somehow colons aren't liked by doc generator.

```suggestion
            intrinsic_matrix: Intrinsic matrix of the camera in row-major format. 
 Shape is (9,).
```

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/ray_caster/patterns/patterns_cfg.py:
```diff
@@ -96,6 +96,53 @@ class PinholeCameraPatternCfg(PatternBaseCfg):
+        width: int,
+        height: int,
+    ) -> PinholeCameraPatternCfg:
+        """Create a PinholeCameraPatternCfg from an intrinsic matrix. The intrinsic matrix is a 3x3 matrix that
+        defines the mapping between the 3D world coordinates and the 2D image. The matrix is defined as:
+
+        .. math::
+            \\begin{bmatrix}
+            f_x & 0 & c_x \\\\
+            0 & f_y & c_y \\\\
+            0 & 0 & 1
+            \\end{bmatrix}
+
```
If you use math symbols, you can use "r" to tell docstrings about that. Then you don't need to do double `\\` anymore. Also the first summary line should be a single sentence. Anything more goes into a paragraph.

```suggestion
        r"""Create an :class:`PinholeCameraPatternCfg` instance from an intrinsic matrix. 
        
        The intrinsic matrix is a 3x3 matrix that defines the mapping between the 3D world coordinates and
        the 2D image. The matrix is defined as:

        .. math::
            \begin{bmatrix}
            f_x & 0 & c_x \\
            0 & f_y & c_y \\
            0 & 0 & 1
            \end{bmatrix}
            
        where :math:`f_x` and :math:`f_y` are the focal length along x and y direction, while :math:`c_x` and :math:`c_y` are the
        principle point offsets along x and y direction respectively.
```

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sim/spawners/sensors/sensors_cfg.py:
```diff
@@ -71,6 +73,72 @@ class PinholeCameraCfg(SpawnerCfg):
+        focal_length: float = 24.0,
+        focus_distance: float = 400.0,
+        f_stop: float = 0.0,
+        horizontal_aperture: float = 20.955,
+        horizontal_aperture_offset: float = 0.0,
+        vertical_aperture_offset: float = 0.0,
```
These are exposed but not used right? 🤔

pascal-roth replied:
did the not implemented error but the arguments are technically set inside the simulator, so I would assume they are used.

pascal-roth replied:
okay they are not used, I will throw a general warning that currently these parameters don't do anything!

pascal-roth replied:
see latest changes

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sim/spawners/sensors/sensors_cfg.py:
```diff
@@ -71,6 +73,72 @@ class PinholeCameraCfg(SpawnerCfg):
+        c_y = intrinsic_matrix[5]
+        # resolve parameters for usd camera
+        horizontal_aperture = width * focal_length / f_x
+        # vertical_aperture = height * focal_length / f_y
```
Maybe we should make a note here as a TODO to check vertical aperture and offset?

Otherwise for now throw a warning that these parameters don't do anything?

pascal-roth replied:
are you sure that they don't do anything?

Mayankm96 commented on source/extensions/omni.isaac.lab/test/sensors/test_camera.py:
```diff
@@ -231,6 +231,57 @@ def test_multi_camera_init(self):
+
+        # play sim
+        self.sim.reset()
+        self.sim.play()
```
Why do you need to do `sim.play()` again? `sim.reset()` does it for you from what I know.

Mayankm96 commented on source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets/stereolabs.py:
```diff
@@ -0,0 +1,80 @@
+
+The following configuration parameters are available:
+
+* :obj:`ZED_X_NARROW_RAYCASTER_CFG`: The ZED X Camera with narrow field of view as RayCasterCameraCfg
+* :obj:`ZED_X_MINI_WIDE_RAYCASTER_CFG`: The ZED X Mini Camera with wide field of view as RayCasterCameraCfg
+* :obj:`ZED_X_NARROW_USD_CFG`: The ZED X Camera with narrow field of view as CameraCfg
+* :obj:`ZED_X_MINI_WIDE_USD_CFG`: The ZED X Mini Camera with wide field of view as CameraCfg
```
```suggestion
* :obj:`ZED_X_NARROW_RAYCASTER_CFG`: The ZED-X Camera with a narrow field of view as an instance of :class:`RayCasterCameraCfg`.
* :obj:`ZED_X_MINI_WIDE_RAYCASTER_CFG`: The ZED-X Mini Camera with a wide field of view as an instance of :class:`RayCasterCameraCfg`.
* :obj:`ZED_X_NARROW_USD_CFG`: The ZED-X Camera with a narrow field of view as an instance of :class:`CameraCfg`.
* :obj:`ZED_X_MINI_WIDE_USD_CFG`: The ZED-X Mini Camera with a wide field of view as an instance of :class:`CameraCfg`.
```

Mayankm96 commented on source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets/stereolabs.py:
```diff
@@ -0,0 +1,80 @@
+    debug_vis=True,
+    max_distance=20,
+    data_types=["distance_to_image_plane"],
+)
```
Missing docstring below the config instance itself?

https://github.com/isaac-sim/IsaacLab/blob/main/source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets/anymal.py#L157-L162

Mayankm96 commented on source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets/stereolabs.py:
```diff
@@ -0,0 +1,80 @@
+        width=960,
+    ),
+    debug_vis=True,
+    max_distance=20,
```
Not sure how well we deal with types but this is float?

```suggestion
    max_distance=20.0,
```

Mayankm96 commented on source/extensions/omni.isaac.lab_assets/omni/isaac/lab_assets/stereolabs.py:
```diff
@@ -0,0 +1,80 @@
+# All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
```
Don't forget to add it to the `__init__.py`

pascal-roth commented:
I enabled that we can set the vertical aperture, mainly because if we change the horizontal aperture in the USD camera case, the vertical one would always stay at the default value of 15. 

Weirdly, the offsets seem to make a difference for both cameras but at a very different scale. I added a test to the raycaster camera that initializes both cameras with the same intrinsic matrix. If the offsets are enabled the values changes as follows:

offsets equal to 0
```
cam_warp_output
tensor([[[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         ...,
         [3.2779, 3.2779, 3.2779,  ..., 3.2778, 3.2778, 3.2778],
         [3.2735, 3.2735, 3.2735,  ..., 3.2734, 3.2734, 3.2734],
         [3.2692, 3.2692, 3.2692,  ..., 3.2691, 3.2691, 3.2691]]],
       device='cuda:0')
cam_usd_output
tensor([[[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         ...,
         [3.2757, 3.2757, 3.2757,  ..., 3.2756, 3.2756, 3.2756],
         [3.2713, 3.2713, 3.2713,  ..., 3.2712, 3.2712, 3.2712],
         [3.2670, 3.2670, 3.2670,  ..., 3.2669, 3.2669, 3.2669]]],
       device='cuda:0')
```

offsets unequal 0
```
cam_warp_output
tensor([[[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         ...,
         [3.2435, 3.2435, 3.2435,  ..., 3.2434, 3.2434, 3.2434],
         [3.2393, 3.2393, 3.2393,  ..., 3.2392, 3.2392, 3.2392],
         [3.2350, 3.2350, 3.2350,  ..., 3.2349, 3.2349, 3.2349]]],
       device='cuda:0')
cam_usd_output
tensor([[[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         ...,
         [3.2748, 3.2748, 3.2748,  ..., 3.2747, 3.2747, 3.2747],
         [3.2704, 3.2704, 3.2704,  ..., 3.2703, 3.2703, 3.2703],
         [3.2661, 3.2661, 3.2661,  ..., 3.2660, 3.2660, 3.2660]]],
       device='cuda:0')
```

@Dhoeller19 and @Mayankm96, do you know how the offsets are handled internally for the USD camera? IMO we handle them correctly for the raycaster case, so that I added warnings now in the USD camera that these values are basically ignored

pascal-roth commented:
@Mayankm96 and @Dhoeller19 ping again about that issue

pascal-roth commented:
@Mayankm96 all issues resolved, can you give it a last check?

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/camera/camera.py:
```diff
@@ -243,6 +245,9 @@ def set_intrinsic_matrices(
             }
+            # TODO: Adjust to handle aperture offsets once supported by omniverse
+            if params["horizontal_aperture_offset"] > 1e-4 or params["vertical_aperture_offset"] > 1e-4:
+                print("[WARNING]: Aperture offsets are not supported by Omniverse cameras. Ignoring offsets.")
```
Use `carb.log_warn`

kellyguo11 commented on source/extensions/omni.isaac.lab/test/sensors/test_camera.py:
```diff
@@ -231,6 +231,56 @@ def test_multi_camera_init(self):
                 for im_data in cam.data.output.to_dict().values():
                     self.assertEqual(im_data.shape, (1, self.camera_cfg.height, self.camera_cfg.width))
 
+    def test_camera_init_intrinsic_matrix(self):
```
does this also apply to the TiledCamera class? if so, should we include a test case in test_tiled_camera.py as well?

pascal-roth replied:
I think that it would be useful if there was a test for it. I personally don't know the TiledCamera. Can you take that over? Should basically be identical to the test here.

pascal-roth commented:
The tiled camera seems to handle the intrinsics wrong. We are running the same function in the background to assign the values to the USDCamera, but the results do not match. I added a test for it in the `test_tiled_camera.py`

 Result of the USD Camera
![cam_test_usd](https://github.com/user-attachments/assets/046c11f0-4218-486b-9360-a4f80d1c66b1)

Result of the Tiled USD Camera
![cam_test_tiled](https://github.com/user-attachments/assets/297fdbf6-db69-44fe-ba81-4718f1f540f0)

Difference between both
![cam_test_diff](https://github.com/user-attachments/assets/d9631489-b52c-4899-b439-cbc5e7a47958)

pascal-roth commented:
also would be great to get some reviews here, as more PR depends on this one @Mayankm96 @jsmith-bdai and @Dhoeller19

kellyguo11 commented:
Thanks for adding the tiled rendering test! Perhaps it may be due to that tiled rendering and multi-render product rendering are using different renderers at the moment. Let's test again with the new tiled rendering.

Also, looks like test_spawn_sensors.py test is failing

pascal-roth commented:
You're correct, the test was failing. Fixed it now.
